README.md
# cloud haskell example

This example sends a sequence of random numbers between nodes using the [cloud-haskell](http://haskell-distributed.github.io/) framework.

## How

The master spawns one process on each node, which it initializes with a sorted list of the process ids.

The first process in the list of process ids elects itself as the leader and starts sending a random number to each of the processes.
The random number is used to select the next leader.
The current leader knows this and can make sure to first send the random number to all other followers in parallel.
When all followers received the random number the current leader can send the number to the next leader.
This way the sequence of random numbers is guaranteed.

When the sending time expires, the master sends a shutdown signal to all processes.
All processes simply store this channel, but the next leader sends its own shutdown signal to a processes.
When a process has received both shutdown signals, it prints out its result and replies to both channels.
Finally the leader also prints out its result and replies to the master's channel.
This way clean shutdown with empty message queues is achieved.

The result is a tuple: `(length msgs, sum (imap (\index msg -> index * msg) msgs))`

## Running

### Requirements

Please install [stack](https://docs.haskellstack.org/en/stable/README/#how-to-install)

### Testing

Testing allows you to run a full cluster as a unit test:

```
stack build
stack test
```

### Running slave

First start up the slaves in the background:

```
stack exec cloud-example-exe -- --port 4445 &
stack exec cloud-example-exe -- --port 4446 &
stack exec cloud-example-exe -- --port 4447 &
```

### Running master

Next start up the master:

```
stack exec cloud-example-exe -- --master --send-for 10 --wait-for 30 --with-seed 123 --discover
```

This master with tell all slaves to send messages for 10 seconds and then wait for 30 seconds for them to cleanly shutdown.
The seed allows the random numbers to be pseudo randomly generated.
The discover flags can be used instead of hard coding the nodes in `./app/Main.hs` in the `hardcodedAddrs` function.

## Problems

  - Sleeping with threadDelay is sleeping too long.
  - Messages always wait for a reply to make sure they arrived at their destination, this could freeze up the whole system.
  - No restart policies with supervision have been added, which should handle issues when nodes or networks give errors.
  - Not only random numbers are being sent, but there is also an initial message and shutdown messages.
  - Require a minimum of at least two slaves.

ChangeLog.md
# Changelog for cloud-example

## Unreleased changes


test/Spec.hs
module Main where

import qualified Test.Tasty as T
import qualified Test.Tasty.HUnit as HUnit
import qualified Exercise

import System.Random (mkStdGen)
import Data.ByteString.Char8 (pack)

import Control.Distributed.Process.Node (initRemoteTable)
import Control.Distributed.Process (NodeId(..))
import Network.Transport (EndPointAddress(..))
import Control.Distributed.Process.Backend.SimpleLocalnet (initializeBackend, startMaster, startSlave)

main :: IO ()
main = T.defaultMain $ T.testGroup "Example" [
        T.testGroup "result" [
            HUnit.testCase "empty" $ HUnit.assertEqual "" (0, 0.0) (Exercise.result [])
            , HUnit.testCase "three items" $ HUnit.assertEqual "" (3, 1*0.3 + 2*0.5 + 3*1.2) (Exercise.result [1.2, 0.5, 0.3])
        ]
        , T.testGroup "select leader" [
            HUnit.testCase "2" $ HUnit.assertEqual "" (1, [2]) (Exercise.selectLeader 0 [1,2])
            , HUnit.testCase "1" $ HUnit.assertEqual "" (1, []) (Exercise.selectLeader 0 [1])
            , HUnit.testCase "4" $ HUnit.assertEqual "" (1, [2,3,4]) (Exercise.selectLeader 0 [1,2,3,4])
            , HUnit.testCase "4 with number 0.33" $ HUnit.assertEqual "" (2, [1,3,4]) (Exercise.selectLeader 0.33 [1,2,3,4])
            , HUnit.testCase "4 with number 0.99" $ HUnit.assertEqual "" (4, [1,2,3]) (Exercise.selectLeader 0.99 [1,2,3,4])
        ]
        , T.testGroup "full program" [
            HUnit.testCase "2 nodes" $ run "4442" (replicate 2 "127.0.0.1:4442:0")
            , HUnit.testCase "3 nodes" $ run "4443" (replicate 3 "127.0.0.1:4443:0")
            , HUnit.testCase "10 nodes" $ run "4440" (replicate 10 "127.0.0.1:4440:0")
        ]
    ]

run :: String -> [String] -> IO ()
run port addrs = do
    backend <- initializeBackend "127.0.0.1" port (Exercise.remoteTable initRemoteTable)
    let r = mkStdGen 0
        nodes = map (NodeId . EndPointAddress . pack) addrs
    startMaster backend (\_ -> Exercise.master backend 1 10 r nodes)


src/Lib.hs
{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE TemplateHaskell #-}

module Lib
    ( remoteTable
    , master
    ) where

import GHC.Generics (Generic)
import Data.Typeable (Typeable)
import Data.Binary
import Data.List (sort)

import System.Environment (getArgs)
import Control.Monad (forM, forM_, when)

import Text.Printf (printf)
import Data.ByteString.Char8 (pack)

import System.Random (RandomGen, random, mkStdGen, StdGen, randoms)
import Control.Distributed.Process
import Control.Concurrent (threadDelay)
import qualified Network.Transport as NT
import Control.Distributed.Process.Closure

import Control.Distributed.Process.Backend.SimpleLocalnet

type ReplyChan = SendPort ProcessId

data Msg = ReplyMsg {
        content :: Content
        , reply :: ReplyChan
    }
    deriving (Typeable, Generic)

data Content = Init {
        ps :: [ProcessId]
    }
    | Number {
        num :: Double
    }
    | DoneFromMaster
    | DoneFromLeader
    deriving (Show, Typeable, Generic)

instance Binary Content
instance Binary Msg

data ShutdownState = Shutdown {
    masterChan :: Maybe ReplyChan
    , leaderChan :: Maybe ReplyChan
}

setMasterChan :: ShutdownState -> ReplyChan -> ShutdownState
setMasterChan (Shutdown _ l) m = Shutdown (Just m) l

setLeaderChan :: ShutdownState -> ReplyChan -> ShutdownState
setLeaderChan (Shutdown m _) l = Shutdown m (Just l)

nodes :: [NodeId]
nodes = map (NodeId . NT.EndPointAddress . pack) ["127.0.0.1:4445:0", "127.0.0.1:4446:0"]

-- toIndex converts the randomly generated number double [0,1) to an index in the list.
toIndex :: [a] -> Double -> Int
toIndex ps d = truncate $ d * fromIntegral (length ps)

-- isLeader checks whether the random number is accosiated with the given item in the list.
isLeader :: (Eq a) => Double -> [a] -> a -> Bool
isLeader n ps p = p == ps !! toIndex ps n

-- selectLeader returns the leader by removing it from the list and returning both the leader and the rest of the list.
selectLeader :: (Eq a) => Double -> [a] -> (a, [a])
selectLeader n ps = 
    let index = toIndex ps n
        (h, t) = splitAt index ps
    in (head t, h ++ tail t)

sendToAll :: Content -> [ProcessId] -> Process [ReceivePort ProcessId]
sendToAll msg ps = forM ps $ \pid -> do
    (sendport,recvport) <- newChan
    send pid (ReplyMsg msg sendport)
    return recvport

waitForAll :: ReceivePort ProcessId -> [ProcessId] -> Process () 
waitForAll _ [] = return ()
waitForAll port ps = do
    pid <- receiveChan port
    waitForAll port (filter (/= pid) ps)

sendAndWait :: Content -> [ProcessId] -> Process ()
sendAndWait msg ps = do
    say $ printf "sending %s to all %s" (show msg) (show ps)
    ports <- sendToAll msg ps
    oneport <- mergePortsBiased ports
    waitForAll oneport ps

-- rand randomly generates a random number that won't reselect the current node as the leader.
rand :: [ProcessId] -> ProcessId -> StdGen -> (Double, StdGen)
rand ps p g = 
    let (r, g') = random g
        index = toIndex ps r
    in if p == (ps !! index)
        then rand ps p g'
        else (r, g')

initNumberNode :: Int -> Process () 
initNumberNode seed = do
    ReplyMsg msg@(Init ps) okChan <- expect
    p <- getSelfPid
    let leader = isLeader 0 ps p
        g = mkStdGen seed
    say $ printf "message received: %s <- %s, leader=%s" (show p) (show msg) (show leader)
    sendChan okChan p
    if leader
        then do 
            let (r, g') = rand ps p g
            propogateNumber ps r
            numberNode g' (Shutdown Nothing Nothing) ps
        else numberNode g (Shutdown Nothing Nothing) ps
    

numberNode :: StdGen -> ShutdownState -> [ProcessId] -> Process ()
numberNode g shutdown ps = do
    p <- getSelfPid
    ReplyMsg msg okChan <- expect
    say $ printf "message received: %s <- %s" (show p) (show msg)
    case msg of
        (Number r) -> do 
            let leader = isLeader r ps p
            say $ printf "%s, leader = %s" (show p) (show leader)
            sendChan okChan p
            if leader
            then case masterChan shutdown of
                Nothing -> do
                    let (r, g') = rand ps p g
                    propogateNumber ps r
                    numberNode g' shutdown ps
                (Just s) -> do
                    sendChan s p
                    say $ printf "leader %s is starting shutdown" (show p)
                    sendAndWait DoneFromLeader ps
                    say $ printf "leader %s is done" (show p)
            else numberNode g shutdown ps
        DoneFromMaster -> nodeShutdown g (setMasterChan shutdown okChan) ps
        DoneFromLeader -> nodeShutdown g (setLeaderChan shutdown okChan) ps

nodeShutdown :: StdGen -> ShutdownState -> [ProcessId] -> Process ()
nodeShutdown g shutdown ps = do
    mypid <- getSelfPid
    case shutdown of
        (Shutdown (Just m) (Just l)) -> do
            say $ printf "shutting down: %s" (show mypid)
            sendChan l mypid
            sendChan m mypid
        _ -> numberNode g shutdown ps

propogateNumber :: [ProcessId] -> Double -> Process ()
propogateNumber ps r = do
    mypid <- getSelfPid
    let msg = Number r
        (leader, followers) = selectLeader r ps
        followersWithoutMe = filter (/= mypid) followers
    say $ printf "sending message: %s -(%s)> %s" (show leader) (show msg) (show followersWithoutMe)
    sendAndWait msg followersWithoutMe
    sendAndWait msg [leader]

remotable ['initNumberNode]

remoteTable :: RemoteTable -> RemoteTable
remoteTable = __remoteTable

spawnAll :: [(Int, NodeId)] -> Process [ProcessId]
spawnAll peers = forM peers $ \(seed, nid) -> do
    say $ printf "spawning on %s" (show nid)
    spawn nid $ $(mkClosure 'initNumberNode) seed

master :: Backend -> Int -> Int -> StdGen -> [NodeId] -> Process () 
master backend sendFor waitFor r peers = do
    let sendForSeconds = sendFor * 1000 * 1000
        waitForSecodds = waitFor * 1000 * 1000
        seeds = take (length peers) $ randoms r
        seeded = zip seeds (sort peers)

    ps <- spawnAll seeded
    refs <- mapM_ monitor ps

    let msg = Init ps
        (leader, followers) = selectLeader 0 ps

    say $ printf "master: leader: %s, followers: %s" (show leader) (show followers)
    sendAndWait msg followers
    sendAndWait msg [leader]

    say "master: init complete"
    liftIO $ threadDelay sendForSeconds

    say "master: starting shutdown"
    sendAndWait DoneFromMaster ps

    say "master: successful shutdown"
    terminateAllSlaves backend


src/Exercise.hs
{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE TemplateHaskell #-}

module Exercise
    ( remoteTable
    , master
    -- * Internal functions
    -- | These functions are exposed for testing purposes.
    , result
    , selectLeader
    ) where

import GHC.Generics (Generic)
import Data.Typeable (Typeable)
import Data.Binary (Binary)

import Data.List (sort)
import Control.Monad (forM, when)
import Text.Printf (printf)
import System.Random (random, mkStdGen, RandomGen, randoms, StdGen)

import Control.Distributed.Process (
    ProcessId, SendPort, ReceivePort, Process, RemoteTable, NodeId
    , say, send, newChan, receiveChan, sendChan, mergePortsBiased, expect, getSelfPid, spawn, monitor, liftIO, spawnLocal)
import Control.Concurrent (threadDelay, newEmptyMVar, putMVar, forkIO)
import Control.Distributed.Process.Closure (mkClosure, remotable)
import Control.Distributed.Process.Backend.SimpleLocalnet (Backend, terminateAllSlaves)

type AckChan = SendPort (ProcessId, [Double])

data Msg = AckMsg {
        content :: Content
        , ackChan :: AckChan
    }
    deriving (Typeable, Generic)

data Content = Init {
        ps :: [ProcessId]
    }
    | Number {
        num :: Double
    }
    | DoneFromMaster
    | DoneFromLeader
    deriving (Show, Typeable, Generic)

instance Binary Content
instance Binary Msg

data ShutdownState = Shutdown {
    masterChan :: Maybe AckChan
    , leaderChan :: Maybe AckChan
}

setMasterChan :: ShutdownState -> AckChan -> ShutdownState
setMasterChan (Shutdown _ l) m = Shutdown (Just m) l

setLeaderChan :: ShutdownState -> AckChan -> ShutdownState
setLeaderChan (Shutdown m _) l = Shutdown m (Just l)

-- | toIndex converts the randomly generated number double [0,1) to an index in the list.
toIndex :: [a] -> Double -> Int
toIndex ps d = truncate $ d * fromIntegral (length ps)

-- | isLeader checks whether the random number is accosiated with the given item in the list.
isLeader :: Double -> [ProcessId] -> Process Bool
isLeader n ps = do
    self <- getSelfPid
    let l = self == ps !! toIndex ps n
    say $ printf "leader = %s" (show l)
    return l

-- | selectLeader returns the leader by removing it from the list and returning both the leader and the rest of the list.
selectLeader :: (Eq a) => Double -> [a] -> (a, [a])
selectLeader r ps = 
    let index = toIndex ps r
        (h, t) = splitAt index ps
    in (head t, h ++ tail t)

sendToAll :: Content -> [ProcessId] -> Process [ReceivePort (ProcessId, [Double])]
sendToAll msg ps = forM ps $ \pid -> do
    (sendport,recvport) <- newChan
    send pid (AckMsg msg sendport)
    return recvport

-- | waitForAll waits to receive an ack from each process, but stops immediately if the process it receives an ack from is itself.
-- This allows us to short circuit the wait loop, by sending a message from ourselves.
waitForAll :: ReceivePort (ProcessId, [Double]) -> [ProcessId] -> Process [[Double]] 
waitForAll _ [] = do
    say "all acks received"
    return []
waitForAll port ps = do
    (pid, rs) <- receiveChan port
    self <- getSelfPid
    if pid == self
        then do
            say "short circuited acks"
            return []
        else do rss <- waitForAll port (filter (/= pid) ps)
                return (rs:rss)

-- | sendAndWait sends content to all processes (excluding self) and waits for acks
sendAndWait :: Content -> [ProcessId] -> Process [[Double]]
sendAndWait _ [] = return []
sendAndWait msg ps = do
    self <- getSelfPid
    let others = filter (/= self) ps
    say $ printf "sending message: %s -> %s" (show msg) (show others)
    ports <- sendToAll msg others
    oneport <- mergePortsBiased ports
    waitForAll oneport others

data RandomProcess g = RandomProcess [ProcessId] ProcessId g

newRandomProcess :: [ProcessId] -> Int -> Process (RandomProcess StdGen)
newRandomProcess ps seed = do
    self <- getSelfPid
    return $ RandomProcess ps self (mkStdGen seed)

-- | rand randomly generates a random number that won't reselect the current node as the leader.
rand :: (RandomGen g) => RandomProcess g -> (Double, RandomProcess g)
rand r@(RandomProcess ps self g) = 
    let (r, g') = random g
        index = toIndex ps r
    in if self == (ps !! index)
        then rand (RandomProcess ps self g')
        else (r, RandomProcess ps self g')

initNumberNode :: Int -> Process ()
initNumberNode seed = do
    AckMsg msg@(Init ps) okChan <- expect
    say $ printf "message received: <- %s" (show msg)
    ack okChan []
    leader <- isLeader 0 ps
    g <- newRandomProcess ps seed
    if leader then do
        let (r, g') = rand g
        propogateNumber ps r
        numberNode [r] g' (Shutdown Nothing Nothing) ps
    else numberNode [] g (Shutdown Nothing Nothing) ps

numberNode :: (RandomGen g) => [Double] -> RandomProcess g -> ShutdownState -> [ProcessId] -> Process ()
numberNode rs g shutdown ps = do
    AckMsg msg ackChan <- expect
    say $ printf "message received: <- %s" (show msg)
    case msg of
        (Number thisr) -> do 
            leader <- isLeader thisr ps
            let newrs = thisr:rs
            ack ackChan []
            if leader
            then case masterChan shutdown of
                Nothing -> do
                    let (newr, g') = rand g
                    propogateNumber ps newr
                    numberNode (newr:newrs) g' shutdown ps
                (Just s) -> do
                    say "leader is starting shutdown"
                    sendAndWait DoneFromLeader ps
                    say $ printf "result: %s" (show $ result newrs)
                    say "shutting down"
                    ack s newrs
            else numberNode newrs g shutdown ps
        DoneFromMaster -> nodeShutdown rs g (setMasterChan shutdown ackChan) ps
        DoneFromLeader -> nodeShutdown rs g (setLeaderChan shutdown ackChan) ps

ack :: AckChan -> [Double] -> Process ()
ack chan rs = do
    self <- getSelfPid
    sendChan chan (self, rs)

-- | result calculates the tuple to be printed out given the reversed list of messages.
result :: [Double] -> (Int, Double)
result = foldr (\d (size, sum) -> (size+1, sum + d * fromIntegral (size+1))) (0,0.0)

nodeShutdown :: (RandomGen g) => [Double] -> RandomProcess g -> ShutdownState -> [ProcessId] -> Process ()
nodeShutdown rs g shutdown ps = case shutdown of
    (Shutdown (Just m) (Just l)) -> do
        say $ printf "result: %s" (show $ result rs)
        say "shutting down"
        ack l []
        ack m rs
    _ -> numberNode rs g shutdown ps

propogateNumber :: [ProcessId] -> Double -> Process ()
propogateNumber ps r = do
    let msg = Number r
        (leader, followers) = selectLeader r ps
    sendAndWait msg followers
    sendAndWait msg [leader]
    return ()

remotable ['initNumberNode]

remoteTable :: RemoteTable -> RemoteTable
remoteTable = __remoteTable

spawnAll :: [(Int, NodeId)] -> Process [ProcessId]
spawnAll peers = forM peers $ \(seed, nid) -> do
    say $ printf "spawning on %s" (show nid)
    spawn nid $ $(mkClosure 'initNumberNode) seed

master :: (RandomGen g) => Backend -> Int -> Int -> g -> [NodeId] -> Process ()
master _ _ _ _ [] = return ()
master _ _ _ _ [p] = return ()
master backend sendFor waitFor r peers = do

    let sendForSecs = sendFor * 1000 * 1000
        waitForSecs = waitFor * 1000 * 1000
        seeds = take (length peers) $ randoms r
        seeded = zip seeds (sort peers)

    ps <- spawnAll seeded
    refs <- mapM_ monitor ps

    let msg = Init ps
        (leader, followers) = selectLeader 0 ps

    say $ printf "master: leader: %s, followers: %s" (show leader) (show followers)
    sendAndWait msg followers
    sendAndWait msg [leader]

    say "master: init complete"
    liftIO $ threadDelay sendForSecs

    say "master: starting shutdown"
    resChan <- sendToAll DoneFromMaster ps
    (sendCancel, recvCancel) <- newChan

    spawnLocal $ do
        liftIO $ threadDelay waitForSecs
        say "master: canceling shutdown"
        ack sendCancel []

    oneResChan <- mergePortsBiased (recvCancel:resChan)
    rs <- waitForAll oneResChan ps

    terminateAllSlaves backend

    when (length rs /= length ps) $ error "did not recceive all results"
    when (any (/= head rs) rs) $ error "not all results are equal"

    say "master: successful shutdown"

app/Main.hs
module Main where

import qualified Exercise

import Options.Applicative
import Data.Semigroup ((<>))

import System.Random (mkStdGen)
import Data.ByteString.Char8 (pack)

import Control.Distributed.Process.Node (initRemoteTable)
import Control.Distributed.Process (NodeId(..))
import Network.Transport (EndPointAddress(..))
import Control.Distributed.Process.Backend.SimpleLocalnet (initializeBackend, startMaster, startSlave)

data Flags = Flags
    { sendFor    :: Int
    , waitFor    :: Int
    , seed       :: Int 
    , master     :: Bool
    , host       :: String
    , port       :: String
    , discover   :: Bool
    }
  deriving (Show)

flags :: Parser Flags
flags = Flags
  <$> option auto
    ( long "send-for" <> help "denotes how many seconds does the system send messages" <> showDefault <> value 1 <> metavar "SECONDS")
  <*> option auto
    ( long "wait-for" <> help "denotes the length of the grace period in seconds" <> showDefault <> value 10 <> metavar "SECONDS")
  <*> option auto
    ( long "with-seed" <> help "How enthusiastically to greet" <> showDefault <> value 1 <> metavar "INT")
  <*> switch
    ( long "master" <> help "Is this the master node or slave" )
  <*> strOption
    ( long "host" <> help "host address" <> showDefault <> value "127.0.0.1")
  <*> strOption
    ( long "port" <> help "host port" <> showDefault <> value "4444")
  <*> switch
    ( long "discover" <> help "discover nodes rather than hard coding them" )

-- | hardcodedAddrs can be modified to reflect the slaves that have been started up
hardcodedAddrs :: [String]
hardcodedAddrs = ["127.0.0.1:4445:0", "127.0.0.1:4446:0", "127.0.0.1:4447:0"]

main :: IO ()
main = do 
  flags <- execParser (info (flags <**> helper) fullDesc)
  backend <- initializeBackend (host flags) (port flags) (Exercise.remoteTable initRemoteTable)
  print flags
  if master flags
    then startMaster backend (\discoveredNodes -> 
          Exercise.master backend
          (sendFor flags)
          (waitFor flags)
          (mkStdGen (seed flags)) 
          (if discover flags then discoveredNodes else map (NodeId . EndPointAddress . pack) hardcodedAddrs))
    else startSlave backend
